### PR TITLE
Support opaque type as kernel argument

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stacked on https://github.com/NVIDIA/Fuser/pull/758

This PR extends `Opaque` and `OpaqueType` to store necessary information for using an opaque type as kernel argument.

### CODE REVIEW

File: test/test_gpu3.cpp [hunk 1]
```diff
@@ -9892,6 +9893,35 @@ TEST_F(NVFuserTest, UnswitchPredicateIssueRepro681_CUDA) {
   testValidate(&fusion, outputs, aten_inputs, {ref}, __LINE__, __FILE__);
 }
 
+TEST_F(NVFuserTest, OpaqueTupleAsComplex) {
```
Comment (zasdfgbnm) [inline test/test_gpu3.cpp:9896]: generated kernel:
```CUDA
__global__ void kernel1(Tuple<float, float> var0, Tensor<std::complex<float>, 1, 1> T0) {
  std::complex<float> c1;
  c1 = std::bit_cast<std::complex<float>>(var0);
  T0[0] = c1;
}
```

**Comment by zasdfgbnm** [comment]
!build

File: csrc/executor_kernel_arg.h [hunk 2]
```diff
@@ -28,9 +28,6 @@ namespace nvfuser {
 //! compilation, we are not unnecessarily holding memory that is not needed.
 class TORCH_CUDA_CU_API KernelArgumentHolder {
  public:
-  //! create KernelArgumentHolder from c10 inputs. Note that we we not taking
-  //! the ownership of the memory from the original inputs, but just recording
-  //! its meta data for kernel execution/compilation.
```
Comment (naoyam) [inline csrc/executor_kernel_arg.h:33]: Why is this deleted?

File: test/test_gpu3.cpp [hunk 3]
```diff
@@ -9893,6 +9894,35 @@ TEST_F(NVFuserTest, UnswitchPredicateIssueRepro681_CUDA) {
...
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  DataType dtype =
+      OpaqueType::make<std::array<float, 2>>("Tuple<float, float>");
+
+  auto tuple = IrBuilder::create<Val>(DataType(dtype));
```
Comment (naoyam) [inline test/test_gpu3.cpp:9904]: Why `DataType(dtype)` rather than just `dtype`?

File: csrc/executor_kernel_arg.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.h:33]: I think this is no longer true. We are taking the ownership, because we keep an instance of `at::Tensor`.

File: test/test_gpu3.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_gpu3.cpp]: Good catch! removed.

**Review by naoyam — APPROVED** [review]
LGTM
